jgiven-tests/src/test/java/com/tngtech/jgiven/impl/ScenarioExecutorTest.java

```diff
@@ -1,88 +1,136 @@
-        String test = buildString();
+    @Test
+    public void set_last_executed_stage_sets_to_the_given_object_if_not_null() {
+        ScenarioExecutor simpleScenarioExecutorObject = new ScenarioExecutor();
+        StageNameWrapper stageNameWrapper = new StageNameWrapper("GIVEN");
 
-        public String buildString() {
-            return "testString";
-        }
+        simpleScenarioExecutorObject.setLastExecutedStageNameWrapper(stageNameWrapper);
 
-        public void the_method_does_not_appear_in_the_report( @Hidden ScenarioCaseModel scenarioCaseModel ) {
-            StepModel stepModel = scenarioCaseModel.getFirstStep();
-            assertThat( stepModel.getWords().get( 1 ).getValue() )
-                .isNotEqualTo( "buildString" )
-                .isEqualTo( "some stage with method called during construction" );
-        }
+        assertThat(simpleScenarioExecutorObject.getLastExecutedStageNameWrapper()).isSameAs(stageNameWrapper);
+    }
 
-        public void some_stage_with_method_called_during_construction() {}
+    @Test
+    public void set_last_executed_stage_does_not_set_to_the_given_object_if_null() {
+        ScenarioExecutor simpleScenarioExecutorObject = new ScenarioExecutor();
+        StageNameWrapper stageNameWrapper = new StageNameWrapper("GIVEN");
+        simpleScenarioExecutorObject.setLastExecutedStageNameWrapper(stageNameWrapper);
 
-        public void a_step_method_with_a_primitive_$_array_$_as_parameter( String type, Object array ) {}
+        simpleScenarioExecutorObject.setLastExecutedStageNameWrapper(null);
 
-        public void the_scenario_is_executed() {}
+        assertThat(simpleScenarioExecutorObject.getLastExecutedStageNameWrapper()).isSameAs(stageNameWrapper);
+    }
 
-        public void no_exception_is_thrown() {
-            assertThat( true ).as( "no exception is thrown" ).isTrue();
-        }
+    @Test
+    public void get_stage_name_wrapper_returns_a_new_stage_name_if_not_null() {
+        ScenarioExecutor simpleScenarioExecutorObject = new ScenarioExecutor();
+        StageNameWrapper stageNameWrapper = new StageNameWrapper("GIVEN");
 
-        public void a_step_failed() {
-            assertThat( false ).isTrue();
-        }
+        doReturn(stageNameWrapper).when((StageNameInternal) mockedStageObject).__jgiven_getStageNameWrapper();
 
-        public void a_NotImplementedYet_annotated_step() {}
+        assertThat(simpleScenarioExecutorObject
+                .getStageNameWrapper(mockedStageObject, null)).isSameAs(stageNameWrapper);
+    }
 
-        public void the_next_step_is_skipped() {}
+    @Test
+    public void get_stage_name_wrapper_returns_the_last_executed_wrapper_if_given_wrapper_null() {
+        ScenarioExecutor simpleScenarioExecutorObject = new ScenarioExecutor();
+        StageNameWrapper stageNameWrapper = new StageNameWrapper("GIVEN");
 
-    }
+        doReturn(null).when((StageNameInternal) mockedStageObject).__jgiven_getStageNameWrapper();
```

Comment:
doesn't line 95 contradict ln 93?

Reply:
The stage name is not set inside mockedObject. What this test is trying to verify is that if the current object we receive as a stage doesn't have a name set to GIVEN/WHEN/THEN (thus having it null), then the last known stage we are executing is not changed. In the case of given().a_simple_test(), given() will set the current stage we are executing to GIVEN, but a_simple_test() has the internal name set to null and we don't want to change the current stage we are executing to that.